### PR TITLE
Add proper handling for invisible widgets.

### LINKED ISSUE #1094: Switching the visibility of charts does not work
I have added a chart in my sitemap, which can be switched "on" with different periods and also can be switched off (by setting all visibilities to falsd), to save time when loading the frame. When all visibilities are "false" (so that no chart is visible) and now one of the chart items is set to visible, no chart is beeing displayed. To make the chart visible, it's necessary to reload the page. The same vice versa, when make the chart invisible the area isn't released.
  
Here is the definition of my chart in the sitemap:
```
Frame {
   Switch item=TemperaturChartPeriod label="Temperaturverlauf" mappings=[0="4 Hours", 1="Day", 2="Week", 3="aus"]                
   Chart item=Wassersensor_1_Temperature period=4h refresh=60   visibility=[TemperaturChartPeriod==0]
   Chart item=Wassersensor_1_Temperature period=D  refresh=600  visibility=[TemperaturChartPeriod==1, TemperaturChartPeriod=="Uninitialized"]
   Chart item=Wassersensor_1_Temperature period=W  refresh=6000 visibility=[TemperaturChartPeriod==2]
}
```

Before updating to version 2.4 (from November 6th) everything was fine (btw. same behavior in the current Beta Version 2.4.1). After reloading the page with "swipe down" the visibility of the chart can be toggled, but when leaving and re-entering the page, the behavior is back.

### PR DESCRIPTION
**Pull request by maniac103**
Track visibility changes via SSE events, and add a fallback for older
servers which triggers a page reload if an unknown widget becomes
visible.

Fixes #1094

### CODE REVIEW

**Comment by maniac103** [comment]
This requires https://github.com/eclipse/smarthome/pull/6856 ... if that one isn't accepted, I'll refactor this PR to only include the reload workaround.

**Comment by mueller-ma** [comment]
I tested this with an official openHAB build and the workaround does its job.
If you're ESH PR gets merged, #1062 will re-appear, I guess.

**Comment by maniac103** [comment]
> If you're ESH PR gets merged, #1062 will re-appear, I guess.

That's a good point, will fix this.

File: mobile/src/main/java/org/openhab/habdroid/ui/activity/PageConnectionHolderFragment.java [hunk 1]
```diff
@@ -278,11 +286,19 @@ private void load() {
             headers.put("X-Atmosphere-tracking-id",
                     mAtmosphereTrackingId != null ? mAtmosphereTrackingId : "0");
 
+            String url = mUrl;
+            if (mCallback.serverCanReturnHiddenWidgets()) {
+                url = Uri.parse(url)
+                        .buildUpon()
+                        .appendQueryParameter("includeHidden", "true")
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/ui/activity/PageConnectionHolderFragment.java]: Usually the server ignores unknown parameters, so you could always add `includeHidden`.

**Comment by maniac103** [comment]
Split off the short-term fix to https://github.com/openhab/openhab-android/pull/1181

**Comment by maniac103** [comment]
https://github.com/openhab/openhab-core/pull/499 is merged, so once the server side settles a bit I'll update this PR.

File: mobile/src/main/java/org/openhab/habdroid/ui/activity/PageConnectionHolderFragment.java (hunk 1 above)
Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/ui/activity/PageConnectionHolderFragment.java]: If that's the case, I can probably drop the whole `serverCanReturnHiddenWidgets` infrastructure...

**Comment by maniac103** [comment]
@mueller-ma This should be RTM now.

**Comment by mueller-ma** [comment]
I retriggered the CI job and it succeeded.
For testing I used the following sitemap:
````
Switch item=Weather_Chart_Period label="Chart Period" icon="chart" mappings=[0="H", 1="D", 2="W", 3="N"]
Chart item=Weather_Chart period=h refresh=600000 visibility=[Weather_Chart_Period==0, Weather_Chart_Period=="NULL"]
Chart item=Weather_Chart period=D refresh=3600000 visibility=[Weather_Chart_Period==1]
Chart item=Weather_Chart period=W refresh=3600000 visibility=[Weather_Chart_Period==2]
````
As long as I switch between `H`, `D` and `W`, it works. On `N` the chart should be hidden (and it is in basic ui), but an empty chart is shown instead.

**Comment by maniac103** [comment]
> but an empty chart is shown instead.

What exactly do you mean by that? Asking because it seems fine for me:
- With one of the first three, I get a chart (placeholder, because I don't have set up a persistence service):
Screenshot_1553594190

- With 'None' selected (your 'N'), I get literally nothing as expected:
Screenshot_1553594194

**Comment by mueller-ma** [comment]
I tested this again: When switching very fast between the different periods, I can see for a short period of time an "empty" chart:
chart
From the log I can see the url: http://openhab/chart?items=Weather_Chart_Period&period=D&random=-800749880&dpi=160&theme=white&w=683&h=341. The item in this case is the "Chart period" selector.

**Comment by maniac103** [comment]
@mueller-ma This is a server bug ... it sends wrong SSE events.

The REST API response for the page looks like this (latest snapshot with `includeHidden` parameter):
```json
{
    "id": "0300",
    "title": "Outside Temperature [- °C]",
    "icon": "temperature",
    "link": "http://localhost:8080/rest/sitemaps/demo/0300",
    "parent": {
        "id": "demo",
        "title": "Main Menu",
        "link": "http://localhost:8080/rest/sitemaps/demo/demo",
        "leaf": true,
        "timeout": false
    },
    "leaf": false,
    "timeout": false,
    "widgets": [
        {
            "widgetId": "030000",
            "type": "Frame",
            "visibility": true,
            "label": "",
            "icon": "frame",
            "mappings": [

            ],
            "widgets": [
                {
                    "widgetId": "03000000",
                    "type": "Text",
                    "visibility": true,
                    "label": "Todays Maximum [- °C]",
                    "icon": "temperature",
                    "mappings": [

                    ],
                    "item": {
                        "link": "http://localhost:8080/rest/items/Weather_Temp_Max",
                        "state": "NULL",
                        "stateDescription": {
                            "pattern": "%.1f °C",
                            "readOnly": false,
                            "options": [

                            ]
                        },
                        "editable": false,
                        "type": "Number",
                        "name": "Weather_Temp_Max",
                        "label": "Todays Maximum",
                        "category": "temperature",
                        "tags": [

                        ],
                        "groupNames": [
                            "Weather",
                            "Weather_Chart"
                        ]
                    },
                    "widgets": [

                    ]
                },
                {
                    "widgetId": "03000001",
                    "type": "Text",
                    "visibility": true,
                    "label": "Todays Minimum [- °C]",
                    "icon": "temperature",
                    "mappings": [

                    ],
                    "item": {
                        "link": "http://localhost:8080/rest/items/Weather_Temp_Min",
                        "state": "NULL",
                        "stateDescription": {
                            "pattern": "%.1f °C",
                            "readOnly": false,
                            "options": [

                            ]
                        },
                        "editable": false,
                        "type": "Number",
                        "name": "Weather_Temp_Min",
                        "label": "Todays Minimum",
                        "category": "temperature",
                        "tags": [

                        ],
                        "groupNames": [
                            "Weather",
                            "Weather_Chart"
                        ]
                    },
                    "widgets": [

                    ]
                },
                {
                    "widgetId": "03000002",
                    "type": "Text",
                    "visibility": false,
                    "label": "Last Update [- -]",
                    "icon": "clock",
                    "mappings": [

                    ],
                    "item": {
                        "link": "http://localhost:8080/rest/items/Weather_LastUpdate",
                        "state": "NULL",
                        "stateDescription": {
                            "pattern": "%1$ta %1$tR",
                            "readOnly": false,
                            "options": [

                            ]
                        },
                        "editable": false,
                        "type": "DateTime",
                        "name": "Weather_LastUpdate",
                        "label": "Last Update",
                        "category": "clock",
                        "tags": [

                        ],
                        "groupNames": [

                        ]
                    },
                    "widgets": [

                    ]
                }
            ]
        },
        {
            "widgetId": "030001",
            "type": "Frame",
            "visibility": true,
            "label": "",
            "icon": "frame",
            "mappings": [

            ],
            "widgets": [
                {
                    "widgetId": "03000100",
                    "type": "Switch",
                    "visibility": true,
                    "label": "Chart Period",
                    "icon": "chart",
                    "mappings": [
                        {
                            "command": "0",
                            "label": "Hour"
                        },
                        {
                            "command": "1",
                            "label": "Day"
                        },
                        {
                            "command": "2",
                            "label": "Week"
                        },
                        {
                            "command": "3",
                            "label": "None"
                        }
                    ],
                    "item": {
                        "link": "http://localhost:8080/rest/items/Weather_Chart_Period",
                        "state": "NULL",
                        "editable": false,
                        "type": "Number",
                        "name": "Weather_Chart_Period",
                        "label": "Chart Period",
                        "tags": [

                        ],
                        "groupNames": [

                        ]
                    },
                    "widgets": [

                    ]
                },
                {
                    "widgetId": "03000101",
                    "type": "Chart",
                    "visibility": true,
                    "label": "Weather_Chart",
                    "icon": "chart",
                    "mappings": [

                    ],
                    "refresh": 600000,
                    "period": "h",
                    "item": {
                        "members": [

                        ],
                        "link": "http://localhost:8080/rest/items/Weather_Chart",
                        "state": "NULL",
                        "editable": false,
                        "type": "Group",
                        "name": "Weather_Chart",
                        "tags": [

                        ],
                        "groupNames": [

                        ]
                    },
                    "widgets": [

                    ]
                },
                {
                    "widgetId": "03000102",
                    "type": "Chart",
                    "visibility": false,
                    "label": "Weather_Chart",
                    "icon": "chart",
                    "mappings": [

                    ],
                    "refresh": 3600000,
                    "period": "D",
                    "item": {
                        "members": [

                        ],
                        "link": "http://localhost:8080/rest/items/Weather_Chart",
                        "state": "NULL",
                        "editable": false,
                        "type": "Group",
                        "name": "Weather_Chart",
                        "tags": [

                        ],
                        "groupNames": [

                        ]
                    },
                    "widgets": [

                    ]
                },
                {
                    "widgetId": "03000103",
                    "type": "Chart",
                    "visibility": false,
                    "label": "Weather_Chart",
                    "icon": "chart",
                    "mappings": [

                    ],
                    "refresh": 3600000,
                    "period": "W",
                    "item": {
                        "members": [

                        ],
                        "link": "http://localhost:8080/rest/items/Weather_Chart",
                        "state": "NULL",
                        "editable": false,
                        "type": "Group",
                        "name": "Weather_Chart",
                        "tags": [

                        ],
                        "groupNames": [

                        ]
                    },
                    "widgets": [

                    ]
                }
            ]
        }
    ]
}
```

Then, when toggling the chart period, the following SSE events arrive:
- `04-01 07:42:43.034  6794  6794 D PageConnectionHolderFragment: Got SSE event {"widgetId":"03000102","label":"Weather_Chart","visibility":false,"item":{"state":"0","editable":false,"type":"Number","name":"Weather_Chart_Period","label":"Chart Period","tags":[],"groupNames":[]},"sitemapName":"demo","pageId":"0300"}` -> not OK (visibility toggle as expected, but with wrong item)
- `04-01 07:42:43.035  6794  6794 D PageConnectionHolderFragment: Got SSE event {"widgetId":"03000100","label":"Chart Period","visibility":true,"item":{"state":"0","editable":false,"type":"Number","name":"Weather_Chart_Period","label":"Chart Period","tags":[],"groupNames":[]},"sitemapName":"demo","pageId":"0300"}` -> OK (period change)
- 04-01 07:42:43.039  6794  6794 D PageConnectionHolderFragment: Got SSE event {"widgetId":"03000101","label":"Weather_Chart","visibility":true,"item":{"state":"0","editable":false,"type":"Number","name":"Weather_Chart_Period","label":"Chart Period","tags":[],"groupNames":[]},"sitemapName":"demo","pageId":"0300"}` -> not OK (visibility toggle with wrong item as above)
- `04-01 07:42:43.041  6794  6794 D PageConnectionHolderFragment: Got SSE event {"widgetId":"03000103","label":"Weather_Chart","visibility":false,"item":{"state":"0","editable":false,"type":"Number","name":"Weather_Chart_Period","label":"Chart Period","tags":[],"groupNames":[]},"sitemapName":"demo","pageId":"0300"}` -> not OK (as above)

In short, the server seems to send SSE events for visibility changes with the item of the trigger, not the actual item of the widget.

Guess we should look into this before 2.5.0 gets released, so it's fixed before the server starts understanding the `includeHidden` API parameter ... without that one, the whole page is reloaded (due to events for unknown widgets), which recovers from that situation.

**Comment by maniac103** [comment]
Opened issue report on openhab-core: https://github.com/openhab/openhab-core/issues/690

**Comment by maniac103** [comment]
@mueller-ma Rebased, should be ready for review now, I think.
